mypy/join.py

```diff
@@ -268,6 +270,12 @@ def visit_tuple_type(self, t: TupleType) -> ProperType:
                                       mypy.typeops.tuple_fallback(t))
             assert isinstance(fallback, Instance)
             return TupleType(items, fallback)
+        elif isinstance(self.s, Instance) and self.s.type.fullname == 'builtins.tuple':
```

Comment:
Couple more things:
* What if the instance is a _subclass_ of `tuple`?
* What if we have a named tuple on either side?

Reply:
I pushed some updates to try and handle the subclass case. I'm afraid I don't really understand the subclassing system. I know to look in .type.mro to see the hierarchy of types, but was unclear how to get the original "args" of the `builtins.tuple` type. For example
```python
class A(Tuple[int, ...]):
    pass

class B(A):
    pass

x: B
y: Tuple[int]
```
How do we access the `builtins.int` part when joining?

Reply:
Once this is resolved, I can work on the NamedTuple case as well